examples/fluids/src/petsc_ops.c

```diff
@@ -12,72 +12,146 @@
     }
   }
 
-  PetscCall(PetscNew(op_apply_ctx));
+  PetscCall(PetscNew(ctx));
 
   // Copy PETSc Objects
   if (dm_x) PetscCall(PetscObjectReference((PetscObject)dm_x));
-  (*op_apply_ctx)->dm_x = dm_x;
+  (*ctx)->dm_x = dm_x;
   if (dm_y) PetscCall(PetscObjectReference((PetscObject)dm_y));
-  (*op_apply_ctx)->dm_y = dm_y;
+  (*ctx)->dm_y = dm_y;
 
   if (X_loc) PetscCall(PetscObjectReference((PetscObject)X_loc));
-  (*op_apply_ctx)->X_loc = X_loc;
+  (*ctx)->X_loc = X_loc;
   if (Y_loc) PetscCall(PetscObjectReference((PetscObject)Y_loc));
-  (*op_apply_ctx)->Y_loc = Y_loc;
+  (*ctx)->Y_loc = Y_loc;
 
   // Copy libCEED objects
-  if (x_ceed) CeedVectorReferenceCopy(x_ceed, &(*op_apply_ctx)->x_ceed);
-  if (y_ceed) CeedVectorReferenceCopy(y_ceed, &(*op_apply_ctx)->y_ceed);
-  CeedOperatorReferenceCopy(op_apply, &(*op_apply_ctx)->op);
-  CeedReferenceCopy(ceed, &(*op_apply_ctx)->ceed);
+  if (x_ceed == CEED_VECTOR_NONE) (*ctx)->x_ceed = x_ceed;
+  else if (x_ceed) CeedVectorReferenceCopy(x_ceed, &(*ctx)->x_ceed);
+  else CeedVectorCreate(ceed, x_size, &(*ctx)->x_ceed);
+
+  if (y_ceed == CEED_VECTOR_NONE) (*ctx)->y_ceed = y_ceed;
+  else if (y_ceed) CeedVectorReferenceCopy(y_ceed, &(*ctx)->y_ceed);
+  else CeedVectorCreate(ceed, y_size, &(*ctx)->y_ceed);
+
+  CeedOperatorReferenceCopy(op_apply, &(*ctx)->op);
+  CeedReferenceCopy(ceed, &(*ctx)->ceed);
 
   PetscFunctionReturn(0);
 }
 
-// -----------------------------------------------------------------------------
-// Destroy apply operator context data
-// -----------------------------------------------------------------------------
-PetscErrorCode OperatorApplyContextDestroy(OperatorApplyContext op_apply_ctx) {
+/**
+ * @brief Destroy OperatorApplyContext struct
+ *
+ * @param[in,out] ctx Context to destroy
+ */
+PetscErrorCode OperatorApplyContextDestroy(OperatorApplyContext *ctx) {
   PetscFunctionBeginUser;
 
-  if (!op_apply_ctx) PetscFunctionReturn(0);
+  if (!(*ctx)) PetscFunctionReturn(0);
 
   // Destroy PETSc Objects
-  PetscCall(DMDestroy(&op_apply_ctx->dm_x));
-  PetscCall(DMDestroy(&op_apply_ctx->dm_y));
-  PetscCall(VecDestroy(&op_apply_ctx->X_loc));
-  PetscCall(VecDestroy(&op_apply_ctx->Y_loc));
+  PetscCall(DMDestroy(&(*ctx)->dm_x));
+  PetscCall(DMDestroy(&(*ctx)->dm_y));
+  PetscCall(VecDestroy(&(*ctx)->X_loc));
+  PetscCall(VecDestroy(&(*ctx)->Y_loc));
 
   // Destroy libCEED Objects
-  CeedVectorDestroy(&op_apply_ctx->x_ceed);
-  CeedVectorDestroy(&op_apply_ctx->y_ceed);
-  CeedOperatorDestroy(&op_apply_ctx->op);
-  CeedDestroy(&op_apply_ctx->ceed);
+  if ((*ctx)->x_ceed != CEED_VECTOR_NONE) CeedVectorDestroy(&(*ctx)->x_ceed);
+  if ((*ctx)->y_ceed != CEED_VECTOR_NONE) CeedVectorDestroy(&(*ctx)->y_ceed);
+  CeedOperatorDestroy(&(*ctx)->op);
+  CeedDestroy(&(*ctx)->ceed);
 
-  PetscCall(PetscFree(op_apply_ctx));
+  PetscCall(PetscFree(*ctx));
```

Comment:
Note - `PetscFree` only takes `ctx`, so you can use `OperatorApplyContextDestroy(OperatorApplyContext ctx)`  as the function signature. This is the signature expected by `MatShellSetContextDestroy` and similar

Reply:
> internals of how `MatShellSetContextDestroy` and similar functions are set up doesn't support that.

In what way?

Reply:
https://petsc.org/main/manualpages/Mat/MatShellSetContextDestroy/

See the required function signature here

Reply:
Ohhh, I hastily misread/misinterpreted `MatShellSetContextDestroy` as Ratel's equivalent to `OperatorApplyContextDestroy` for some reason. Yeah, that makes sense. I'm guessing is similarly doesn't make sense to have two different destroy functions

Reply:
Yeah, I wouldn't make separate functions.

We could go into PETSc and change the signature if we need - I don't think anyone else but us is using those functions. I followed the pattern of `PetscFree`  when creating the functions.